### PR TITLE
[clang][flang][mlir] Support -frecord-command-line option

### PR DESCRIPTION
**Pull request by tarunprabhu**
Add support for the -frecord-command-line option that will produce the llvm.commandline metadata which will eventually be saved in the object file. This behavior is also supported in clang. Some refactoring of the code in flang to handle these command line options was carried out. The corresponding -grecord-command-line option which saves the command line in the debug information has not yet been enabled for flang.

--------------------------------------------------------
Notes for reviewers: In #98517, the `tune-cpu` argument was added to the constructor of the lowering bridge object in flang. This PR would have required adding another option to it. It is possible more such arguments may be added in the future. To avoid having to add new arguments each time, the `TargetOptions` and `CodeGenOptions` objects from the frontend are now passed directly through to the lowering bridge constructor. 

This required a change to `bbc` where empty `TargetOptions` and `CodeGenOptions` objects had to be created. 

If others have better suggestions for how to handle this, I am happy to make changes.

[EDIT: Fix typo]

### CODE REVIEW

**Review by Dinistro — APPROVED** [review]
The MLIR part looks good to me, but I cannot speak for clang nor flang 🙂

File: clang/include/clang/Driver/Options.td
Comment (kiranchandramohan) [inline clang/include/clang/Driver/Options.td:1983]: Will the DocBrief section above lead to mentions of Clang in https://flang.llvm.org/docs/FlangCommandLineReference.html?

File: clang/include/clang/Driver/Options.td
Comment (kiranchandramohan) [inline clang/include/clang/Driver/Options.td:1983]: Do you know whether flang's flow is affected by `f_clang_Group`?

**Review by kiranchandramohan — COMMENTED** [review]
Couple of minor questions.

File: clang/include/clang/Driver/Options.td
Comment (tarunprabhu) [reply clang/include/clang/Driver/Options.td:1983]: Nice catch. The text only contains a single reference to clang. I have removed it, so it should apply to both `clang` and `flang` now.

File: clang/include/clang/Driver/Options.td
Comment (tarunprabhu) [reply clang/include/clang/Driver/Options.td:1983]: Could you clarify what you mean by "flang's flow"? 

I don't know exactly what `f_clang_Group` does. I haven't been able to determine what effect it has on things like help messages. For instance `-frecord-command-line` does not appear in either `clang --help`, or `clang --help-hidden`.

Comment (tarunprabhu) [reply clang/include/clang/Driver/Options.td:1991]: @kiranchandramohan, gentle ping

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

File: clang/include/clang/Driver/Options.td
Comment (kiranchandramohan) [reply clang/include/clang/Driver/Options.td:1991]: I mean't whether `f_clang_Group` is processed by flang or not. I see that this is used by some other commands as well. So it is alrite.

Comment (MaskRay) [reply clang/include/clang/Driver/Options.td:1991]: I think f_clang_Group is a legacy group for Clang options not in GCC. f_clang_Group has never gain significant adoption and it's used by any code.

New -f options can just use f_Group.

**Comment by MaskRay** [comment]
Why do we need the two new files?

(added) clang/lib/Driver/ToolChains/CommonUtils.cpp (+76)
(added) clang/lib/Driver/ToolChains/CommonUtils.h (+44)

We have clang/lib/Driver/ToolChains/CommonArgs.cpp and clang/lib/Driver/ToolChain.cpp for shared code.

**Comment by tarunprabhu** [comment]
> Why do we need the two new files?
> 
> We have clang/lib/Driver/ToolChains/CommonArgs.cpp and clang/lib/Driver/ToolChain.cpp for shared code.

Since most of the code in `CommonArgs.cpp` had to do with functions in the `clang::driver::tools` namespace, I wasn't sure that code that is only shared between `Clang.cpp` and `Flang.cpp` ought to go there. The other function that was moved is a utility function to escape arguments and was originally declared `static` in `Clang.cpp`.

If it is ok to use `CommandArgs.cpp` for limited sharing and utilities, I can move the code there.

File: clang/include/clang/Driver/Options.td
Comment (tarunprabhu) [reply clang/include/clang/Driver/Options.td]: I have changed the group to `f_Group` since that does not suggest that the options are exclusive to clang.

**Comment by MaskRay** [comment]
> > Why do we need the two new files?
> > We have clang/lib/Driver/ToolChains/CommonArgs.cpp and clang/lib/Driver/ToolChain.cpp for shared code.
> 
> Since most of the code in `CommonArgs.cpp` had to do with functions in the `clang::driver::tools` namespace, I wasn't sure that code that is only shared between `Clang.cpp` and `Flang.cpp` ought to go there. The other function that was moved is a utility function to escape arguments and was originally declared `static` in `Clang.cpp`.
> 
> If it is ok to use `CommandArgs.cpp` for limited sharing and utilities, I can move the code there.

Adding the shared code to `CommonArgs.cpp` LGTM. Shared functions there use `tools::`, though.

`rg driver:: clang/lib/Driver` reveals what files define `driver::` member functions. `CommonArgs.cpp` doesn't define `driver::`.

**Comment by tarunprabhu** [comment]
Moved code to the `clang::driver::tools` namespace

**Comment by tarunprabhu** [comment]
@MaskRay, If everything looks good to you now, could you approve this? 

I apologize for being pushy, but this will help us get moving on some other things in `flang`. Thanks.

**Review by MaskRay — APPROVED** [review]
LGTM.

As you are renaming functions, you can fix the function naming style 
`tools::function` instead of `tools::Function`.
While the style is largely inconsistent, we usually fix the style while renaming functions.

**Comment by tarunprabhu** [comment]
> As you are renaming functions, you can fix the function naming style `tools::function` instead of `tools::Function`. While the style is largely inconsistent, we usually fix the style while renaming functions.

I will do that and merge. Thanks!

**Comment by DavidSpickett** [comment]
I've had to revert this due to the failures listed above. To reproduce, configure with `-DBUILD_SHARED_LIBS=ON`.

(the other 2 bots are not called shared libs but they do enable this setting too)

```
$ ./bin/bbc
: CommandLine Error: Option 'fdynamic-heap-array' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
Aborted (core dumped)
```

This happens to at least the bbc and tco tools that I notice include a .inc file related to commands. So that might be the issue, but simply removing the include wasn't correct so I leave it to you to figure out.

**Comment by tarunprabhu** [comment]
Thanks for the instructions on how to reproduce. I'll take a look.

**Comment by tarunprabhu** [comment]
The underlying issue is due to `CLOptions.inc` being included in `FrontendActions.cpp` and `bbc.cpp`. The file contained a number of command line options and functions, all of which were declared `static`. This PR added `libflangFrontend` to the list of libraries linked into `bbc` which resulted in duplicate definition errors. The underlying issue can be solved by avoiding including files containing definitions directly in another. #109874 does this. It should be safe to reland this after that is merged.

**Pull request merged**
